File: BOJ/BOJ_11123_양한마리...양두마리....cpp
#include <bits/stdc++.h>
using namespace std;

int T, H, W, cnt;
char graph[101][101];
bool visited[101][101];
int dy[4] = {-1, 1, 0, 0};
int dx[4] = {0, 0, -1, 1};

void init() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
}

void DFS(int y, int x) {
    visited[y][x] = true;
    for (int i = 0; i < 4; i++) {
        int ny = y + dy[i];
        int nx = x+ dx[i];
        if (0 <= ny && ny < H && 0 <= nx && nx < W && graph[ny][nx] == '#' && !visited[ny][nx]) {
            DFS(ny, nx);
        }
    }
}

int main(int argc, char const *argv[]) {
    init();
    cin >> T;
    while(T--) {
        cin >> H >> W;
        cnt = 0;
        memset(visited, 0, sizeof(visited));
        memset(graph, 0, sizeof(graph));

        for (int i = 0; i < H; i++) {
            for (int j = 0; j < W; j++) {
                cin >> graph[i][j];
            }
        }
        for (int i = 0; i < H; i++) {
            for (int j = 0; j < W; j++) {
                if (graph[i][j] == '#' && !visited[i][j]) {
                    DFS(i, j);
                    cnt++;
                }
            }
        }
        cout << cnt << '\n';
    }
    return 0;
}

File: BOJ/BOJ_11123_양한마리...양두마리....py
import sys
sys.setrecursionlimit(10**6)

T = int(input())
H = W = cnt = 0
dy = [-1, 1, 0, 0]
dx = [0, 0, -1, 1]

def dfs(y, x, graph, visited):
    visited[y][x] = 1
    for i in range(4):
        ny, nx = y + dy[i], x + dx[i]
        if 0 <= ny < H and 0 <= nx < W and graph[ny][nx] == '#' and not(visited[ny][nx]):
            dfs(ny, nx, graph, visited)

for t in range(T):
    H, W = map(int, input().split())
    graph = [input() for _ in range(H)]
    visited = [[0] * W for _ in range(H)]
    cnt = 0

    for i in range(H):
        for j in range(W):
            if graph[i][j] == '#' and not(visited[i][j]):
                dfs(i, j, graph, visited)
                cnt += 1

    print(cnt)